azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java

```diff
@@ -1954,6 +1955,10 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
       });
 
       registerErrorsAndWarningsFromValidationReport(resp, ret, reports);
+      // Reload the flow_filter to ensure if this project is added to the file, then the filter is current.
+      ContainerizedDispatchManager containerizedDispatchManager =
+          (ContainerizedDispatchManager) this.executorManagerAdapter;
+      containerizedDispatchManager.getContainerFlowCriteria().reloadFlowFilter();
```

Comment:
In current implementation of reloadFlowFilter(), we will do follows:
1. Reset "Map<String, Set<String>> flows"
2. Parse flow filter file and add flow into flows map.

This can cause similar ordering issue when:
1. Project is being uploaded
2. Flow is being scheduled/executed at the same time.

As a result, flow will be scheduled to containerization b/c flow map is almost empty / slowly added during the reloadFlowFilter(). As a result, I think we should change the logic inside of reloadFlowFilter:
1. Parse flow filter file and add flow into flows map.
2. Replace the current "Map<String, Set<String>> flows"

Reply:
Good suggestion. This avoids race condition. Will update the API.

Reply:
BTW, I feel like the fix I proposed is incorrect. Instead, the ideal way is to add a lock in getDispatchMethod() and reloadFlowFilter() so that only one can be executed at a time. I'm not pretty sure re-pointing the flows map (the previous design) will still cause race condition or not.